doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,487 @@
+  performance cost. It also aids users of the API to understand that this
+  functionality is not always available and to detect when this functionality is
+  available in advance.
+
+### Design
+
+The currently proposed design is as follows:
+
+```c
+typedef struct bio_msg_st {
+    void *data;
+    size_t data_len;
+    BIO_ADDR *peer, *local;
+    uint64_t flags;
+} BIO_MSG;
+
+#define BIO_UNPACK_ERRNO(e)     /*...*/
+#define BIO_IS_ERRNO(e)         /*...*/
+
+ossl_ssize_t BIO_sendmmsg(BIO *b, BIO_MSG *msg, size_t stride,
+                          size_t num_msg, uint64_t flags);
+ossl_ssize_t BIO_recvmmsg(BIO *b, BIO_MSG *msg, size_t stride,
+                          size_t num_msg, uint64_t flags);
+```
+
+The API is used as follows:
+
+- `msg` points to an array of `num_msg` `BIO_MSG` structures.
+
+- Both functions have identical prototypes, and return the number of messages
+  processed in the array. If no messages were sent due to an error, `-1` is
+  returned. If an OS-level socket error occurs, a negative value `v` is
+  returned. The caller should determine that `v` is an OS-level socket error by
+  calling `BIO_IS_ERRNO(v)` and may obtain the OS-level socket error code by
+  calling `BIO_UNPACK_ERRNO(v)`.
+
+- `stride` must be set to `sizeof(BIO_MSG)`.
+
+- `data` points the buffer of data to be sent or to be filled with received
+  data. `data_len` is the size of the buffer in bytes on call. If the
+  given message in the array is processed (i.e., if the return value
+  exceeds the index of that message in the array), `data_len` is updated
+  to the actual amount of data sent or received at return time.
+
+- `flags` in the `BIO_MSG` structure provides per-message flags to
+  the `BIO_sendmmsg` or `BIO_recvmmsg` call. If the given message in the array
+  is processed, `flags` is written with zero or more result flags at return
+  time. The `flags` argument to the call itself provides for global flags
+  affecting all messages in the array. Currently, no per-message or global flags
+  are defined and all of these fields are set to zero on call and on return.
+
+- `peer` and `local` are optional pointers to `BIO_ADDR` structures into
+  which the remote and local addresses are to be filled. If either of these
+  are NULL, the given addressing information is not requested. Local address
+  support may not be available in all circumstances, in which case processing of
+  the message fails. (This means that the function returns the number of
+  messages processed, or -1 if the message in question is the first message.)
+
+  Support for `local` must be explicitly enabled before use, otherwise
+  attempts to use it fail.
+
+Local address support is enabled as follows:
+
+```c
+int BIO_dgram_set_local_addr_enable(BIO *b, int enable);
+int BIO_dgram_get_local_addr_enable(BIO *b);
+int BIO_dgram_get_local_addr_cap(BIO *b);
+```
+
+`BIO_dgram_get_local_addr_cap()` returns 1 if local address support is enabled.
```

Comment:
What is the difference between `BIO_dgram_get_local_addr_enable` and `BIO_dgram_get_local_addr_cap`?

Reply:
Yeah, I should have said "available" here.